[10.0][ADD] shopinvader_validation_token

acsonefho opened the pull request:
This module add a security layer for the guest and customer service.

Currently a customer can use any email address (even if he is not the owner of it).
With this module, a code/token is sent to the email address and the customer has to fill it to continue the process.

Include https://github.com/shopinvader/odoo-shopinvader/pull/1013

acsonefho committed:
[10.0][ADD] shopinvader_validation_token: add a security code/token sent as a notification (currently email) to ensure the customer is the owner of the given email address

simahawk commented:
nice addition, BUT: can we have it in its own module? This could be used to validate also normal registrations or any other email.
Eg: `shopinvader_validation_token`

Cedric-Pigeon commented on shopinvader_validation_token/README.rst:
```diff
@@ -0,0 +1,69 @@
+
+|badge1| |badge2| |badge3| 
+
+This Shopinvader addon implement a security layer on the guest/customer mode.
```
"implements"

Cedric-Pigeon commented on shopinvader_validation_token/README.rst:
```diff
@@ -0,0 +1,69 @@
+
+This Shopinvader addon implement a security layer on the guest/customer mode.
+To ensure a customer doesn't use the email of somebody else,
+it generates a token/code send using notification and
```
"sent"

Cedric-Pigeon commented on shopinvader_validation_token/README.rst:
```diff
@@ -0,0 +1,69 @@
+This Shopinvader addon implement a security layer on the guest/customer mode.
+To ensure a customer doesn't use the email of somebody else,
+it generates a token/code send using notification and
+the customer have to fill it on the front-side.
```
"has"

Cedric-Pigeon commented on shopinvader_validation_token/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Shopinvader validation token",
+    "summary": """Implement a security layer on the registration and/or
```
"Implements

Cedric-Pigeon commented on shopinvader_validation_token/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+    "summary": """Implement a security layer on the registration and/or
+    guest mode. To ensure a customer doesn't use the email of somebody else,
+    it generates a token/code send using notification and
+    the customer have to fill it on the front-side.""",
```
"has"

Cedric-Pigeon commented on shopinvader_validation_token/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+    the customer have to fill it on the front-side.""",
+    "version": "10.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
no OCA, at least for the time being

Cedric-Pigeon commented on shopinvader_validation_token/data/ir_cron.xml:
```diff
@@ -0,0 +1,18 @@
+        <field eval="True" name="active"/>
+        <field name="user_id" ref="base.user_root"/>
+        <field name="interval_number">1</field>
+        <field name="interval_type">days</field>
```
i would set it more often, 1 time per hour

Cedric-Pigeon commented on shopinvader_validation_token/models/shopinvader_backend.py:
```diff
@@ -0,0 +1,25 @@
+    _inherit = "shopinvader.backend"
+
+    token_validity = fields.Integer(
+        string="Token validity (minutes)", default=60,
```
I would reduce the default to 20

Cedric-Pigeon commented on shopinvader_validation_token/models/shopinvader_security_token.py:
```diff
@@ -0,0 +1,158 @@
+        :param email: str
+        :return: str
+        """
+        return str(uuid4())
```
I would rather random a number on 6 digits from "000000" to "999999"

Cedric-Pigeon commented on shopinvader_validation_token/services/customer_service.py:
```diff
@@ -0,0 +1,83 @@
+from odoo.addons.component.core import Component
+
+
+class ShopinvaderCustomerService(Component):
```
I do not understand this. Why do you impact the create depending on a guest feature?
If you want to allow that it should be possible without activating the guest mode which is not like you did

acsonefho replied:
This feature is not only for guest.
Depending on the configuration/notification, it's available for `guest`, `customer` or both.

As a `guest` service inherits `customer` service, it's applied on both services.

acsonefho commented:
> nice addition, BUT: can we have it in its own module? This could be used to validate also normal registrations or any other email.
> Eg: `shopinvader_validation_token`

I update my PR with your suggestion. Can you review it please?

Thanks!

simahawk commented on shopinvader_validation_token/README.rst:
```diff
@@ -0,0 +1,69 @@
+* By interface:
+
+- Browse into your shopinvader backends;
+- Add a notification named "Security token" (for customer service);
```
not required here but would be nice to see a backport of https://github.com/shopinvader/odoo-shopinvader/tree/13.0/shopinvader_notification_default and a new default template for this token there ;)

simahawk commented on shopinvader_validation_token/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+    "version": "10.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV",
+    "website": "https://acsone.eu/",
```
shoinvader,com or https://github.com/shopinvader

simahawk commented on shopinvader_validation_token/models/shopinvader_security_token.py:
```diff
@@ -0,0 +1,162 @@
+
+    @api.model
+    def _disable_expired_token(self):
+        """
```
https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

simahawk commented on shopinvader_validation_token/models/shopinvader_security_token.py:
```diff
@@ -0,0 +1,162 @@
+            ("expiration_datetime", "<", now),
+        ]
+        shopinvader_tokens = self.sudo().search(domain)
+        if shopinvader_tokens:
```
this `if` is superfluous

simahawk commented on shopinvader_validation_token/models/shopinvader_security_token.py:
```diff
@@ -0,0 +1,162 @@
+            "email": email,
+            "token": token,
+            "expiration_datetime": expiration_datetime,
+            "notification_type": notif_type,
```
instead of binding this to `notification_type` wouldn't be better to have `token_type`?
The notification type can be `service_name:token_type`.

simahawk replied:
We should probably have an independent token configuration instead.
With a new model, with a python snippet that allows to hooks this machinery via configuration in any service.
What is your main goal? Validate emails? Validate important changes?

simahawk replied:
what about this?

simahawk commented on shopinvader_validation_token/tests/common.py:
```diff
@@ -0,0 +1,94 @@
+    def _get_notification_type(self):
+        return ""
+
+    def setUp(self, *args, **kwargs):
```
you can do everything in setUpClass

rousseldenis replied:
@acsonefho Could you attend ?

@simahawk Not the work_on_services part I think

simahawk reviewed (changes requested):
I'm not sure about the implementation. I'll pass by again an rethink about it should work.
Of course, this is not a blocker... If you find consensus, go ahead w/ what you have here ;)

rousseldenis committed:
[10.0][FIX] python-magic fix

acsonefho committed:
[10.0][FIX] shopinvader_product_template_multi_link:
- Fix compute due to changes into #452;
- fix Unit tests;
- Improve loop part to reduce complexity.
Memo: unit tests has been fixed before to improve nested loops.

acsonefho committed:
[IMP] pre-commit

rousseldenis commented:
@simahawk Could you update your review ? Thanks

rousseldenis commented:
@acsonefho Could you check as it seems to include code from other modules ?

acsonefho commented:
> @acsonefho Could you check as it seems to include code from other modules ?

Because this PR include #1013

simahawk commented on shopinvader_validation_token/models/shopinvader_security_token.py:
```diff
@@ -0,0 +1,167 @@
+            notif_type = self._service_notification_map().get(service_name, "")
+            if notif_type:
+                should_trigger = any(
+                    n.notification_type == notif_type
```
I don't get the reason of this logic: the token stuff is enabled only if you have the notification? :thinking:

simahawk commented on shopinvader_validation_token/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,5 @@
+* By interface:
+
+- Browse into your shopinvader backends;
+- Add a notification named "Security token" (for customer service);
```
enabling this via notification seems weird

simahawk reviewed (changes requested):
I have some serious doubt regarding the logic behind, but maybe I'm missing something :man_shrugging:

lmignon commented:
Closing since 10.0 is no more actively maintained.
